### PR TITLE
Convert `Guest` tests to JS

### PR DESCRIPTION
**Pull request by robertknight**
This is the result of an automated conversion by decaffeinate followed
by a manual review and clean-up of the result.

The diff is large so I don't expect a line-by-line comparison of new and old. A skim through the converted code looking for anything obvious amiss would be useful though.

As mentioned in https://github.com/hypothesis/client/pull/2523, there are currently coverage gaps in the tests for the `Guest` class. I haven't addressed those here but will do so in follow-up PRs.

### CODE REVIEW

File: src/annotator/test/guest-test.js [hunk 1]
```diff
@@ -0,0 +1,869 @@
...
+      guest = createGuest();
+      guest.plugins.PDF = {
+        uri() {
+          return 'urn:x-pdf:001122';
+        },
+        getMetadata: sandbox.stub(),
+      };
+    });
+
+    it('preserves the components of the URI other than the fragment', () => {
```
Comment (lyzadanger) [inline src/annotator/test/guest-test.js:526]: It's intriguing that this is the responsibility of `guest`...

File: src/annotator/test/guest-test.js [hunk 2]
```diff
@@ -0,0 +1,869 @@
...
+
+      guest.createAnnotation(annotation);
+
+      return timeoutPromise().then(() => {
+        assert.equal(annotation.uri, 'http://example.com/');
+        assert.deepEqual(annotation.document, { title: 'hello' });
+      });
+    });
+
+    it('treats an argument as the annotation object', () => {
```
Comment (lyzadanger) [inline src/annotator/test/guest-test.js:566]: This description makes me go "huh?"

File: src/annotator/test/guest-test.js [hunk 3]
```diff
@@ -0,0 +1,869 @@
...
+      highlighter.highlightRange.returns(highlights);
+      const target = {
+        selector: [{ type: 'TextQuoteSelector', exact: 'hello' }],
+      };
+      return guest
+        .anchor({ target: [target] })
+        .then(anchors => assert.equal(anchors.length, 1));
+    });
+
+    it('adds the anchor to the "anchors" instance property"', () => {
```
Comment (lyzadanger) [inline src/annotator/test/guest-test.js:778]: This feels a little implementation-detail-y...I'm curious what this is attempting to test at a higher level.

**Review by lyzadanger — APPROVED** [review]
Given the abstract nature of this review, my comments are probably mostly annoying/pedantic!

File: src/annotator/test/guest-test.js [hunk 4]
```diff
@@ -0,0 +1,869 @@
...
+
+  const createGuest = (config = {}) => {
+    config = Object.assign({}, guestConfig, config);
+    const element = document.createElement('div');
+    const guest = new Guest(element, config);
+    return guest;
+  };
+
+  beforeEach(() => {
+    sinon.stub(console, 'warn');
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/test/guest-test.js:56]: It turns out this `console.warn` stub is obsolete.

File: src/annotator/test/guest-test.js [hunk 5]
```diff
@@ -0,0 +1,869 @@
...
+
+  describe('plugins', () => {
+    let fakePlugin;
+    let guest;
+
+    beforeEach(() => {
+      FakePlugin.instance = null;
+      guestConfig.pluginClasses.FakePlugin = FakePlugin;
+      guest = createGuest({ FakePlugin: {} });
+      fakePlugin = FakePlugin.instance;
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/test/guest-test.js:118]: `FakePlugin.instance` is a way to get a reference to the most recently constructed instance of the `FakePlugin` class which happens inside the `Guest` constructor.

File: src/annotator/test/guest-test.js [hunk 6]
```diff
@@ -0,0 +1,869 @@
...
+    });
+
+    it('load and "pluginInit" gets called', () => {
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/test/guest-test.js:121]: Yes, indeed. Will tidy up.

File: src/annotator/test/guest-test.js [hunk 7]
```diff
@@ -0,0 +1,869 @@
...
+        );
+        const highlight1 = $(
+          '<span class="hypothesis-highlight-focused"></span>'
+        );
+        const guest = createGuest();
+        guest.anchors = [
+          { annotation: { $tag: 'tag1' }, highlights: highlight0.toArray() },
+          { annotation: { $tag: 'tag2' }, highlights: highlight1.toArray() },
+        ];
+        emitGuestEvent('focusAnnotations', 'ctx', ['tag1']);
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/test/guest-test.js:253]: This looks like a mistake. Other uses of this event don't have this argument and the tests pass without it.

File: src/annotator/test/guest-test.js [hunk 8]
```diff
@@ -0,0 +1,869 @@
...
+          uri: sandbox.stub().returns(window.location.href),
+          getMetadata: sandbox.stub(),
+        };
+      });
+
+      afterEach(() => {
+        sandbox.restore();
+      });
+
+      it('calls the callback with the href and pdf metadata', done => {
```
[the remark this answers is not in the document]
Comment (robertknight) [reply src/annotator/test/guest-test.js:332]: Indeed. It's not immediately obvious to me what is going on here either. I'm going to leave this as an exercise for later.

File: src/annotator/test/guest-test.js (hunk 3 above)
Comment (robertknight) [reply src/annotator/test/guest-test.js:778]: The `anchors` property is currently part of the public interface of `Guest` which `bucket-bar.coffee` and possibly others rely on. The code design could be improved but this test is currently appropriately testing part of the public interface.

**Pull request merged**